product_customer_code/models/product.py

```diff
@@ -0,0 +1,31 @@
+# -*- coding: utf-8 -*-
+# Copyright 2012 Vauxoo - http://www.vauxoo.com
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    product_customer_code_ids = fields.One2many(
+        comodel_name='product.customer.code',
+        inverse_name='product_id',
+        string='Customer Codes',
+        copy=False
+    )
+
+    @api.model
+    def name_search(self, name='', args=None, operator='ilike', limit=100):
+        res = super(ProductProduct, self).name_search(
+            name=name, args=args, operator=operator, limit=limit)
+        if (len(res) < limit):
+            partner_id = self._context.get('partner_id')
+            if partner_id:
+                product_customer_code_obj = self.env['product.customer.code']
+                product_codes = product_customer_code_obj.search([
+                    ('product_code', operator, name),
+                    ('partner_id', '=', partner_id)
+                ], limit=limit)
+                res.extend(product_codes.mapped('product_id').name_get())
```

Comment:
Is this what we want?  The behaviour before this commit was the customer code.  Now I assume it is the [product_code] product name if the display_default_code context key is set.  

Would something like 
```python
product_name_get = product_codes.mapped('product_id').with_context(display_default_code=False).name_get()
if self._context.get('display_default_code'):
    product_name_get = [(pk, '[%s] %s' % product_codes.filtered(lambda r: r.product_id = pk).product_code
res.extend(product_name_get)
```
or adjusting name_get.  Its not great code but hopefully it gives an idea what I mean.  Its just from a user perspective I think this will be confusing.  Type a product code and get a different one back.  Note I've not looked at product_customer_code_sale so maybe its handled there better.